### PR TITLE
Oak-9126: TestFramework: create tests for both lucene and elastic

### PR DESCRIPTION
**Pull request by tihom88**

### CODE REVIEW

File: oak-jcr/src/test/java/org/apache/jackrabbit/oak/query/AbstractJcrTest.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package org.apache.jackrabbit.oak.query;
```
Comment (fabriziofortino) [inline oak-jcr/src/test/java/org/apache/jackrabbit/oak/query/AbstractJcrTest.java:19]: add license

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/LuceneIndexOptions.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package org.apache.jackrabbit.oak.plugins.index;
```
Comment (fabriziofortino) [inline oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/LuceneIndexOptions.java:19]: add license

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/FacetTest1.java [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+import static org.apache.jackrabbit.oak.plugins.index.search.FulltextIndexConstants.PROP_SECURE_FACETS;
+import static org.apache.jackrabbit.oak.plugins.index.search.FulltextIndexConstants.PROP_SECURE_FACETS_VALUE_INSECURE;
+import static org.apache.jackrabbit.oak.plugins.index.search.FulltextIndexConstants.PROP_SECURE_FACETS_VALUE_STATISTICAL;
+import static org.apache.jackrabbit.oak.plugins.index.search.FulltextIndexConstants.PROP_STATISTICAL_FACET_SAMPLE_SIZE;
+import static org.apache.jackrabbit.oak.plugins.index.search.FulltextIndexConstants.STATISTICAL_FACET_SAMPLE_SIZE_DEFAULT;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertTrue;
+
+public abstract class FacetTest1 extends AbstractJcrTest {
```
Comment (fabriziofortino) [inline oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/FacetTest1.java]: Why do we need the `1` in the name? This is a question also for all the other tests with the same suffix

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/TestRepository.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package org.apache.jackrabbit.oak.plugins.index;
```
Comment (fabriziofortino) [inline oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/TestRepository.java:19]: add license

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/TestRepositoryBuilder.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
+package org.apache.jackrabbit.oak.plugins.index;
```
Comment (fabriziofortino) [inline oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/TestRepositoryBuilder.java:19]: add license

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/TestRepository.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+import org.apache.jackrabbit.oak.Oak;
+import org.apache.jackrabbit.oak.spi.state.NodeStore;
+
+public class TestRepository {
+
+    private AsyncIndexUpdate asyncIndexUpdate;
```
Comment (fabriziofortino) [inline oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/TestRepository.java:26]: is this needed? we just set this at line 39 but we never access it again. Should we read it in `isAsync` at line 34?

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/TestRepositoryBuilder.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    protected TrackingCorruptIndexHandler trackingCorruptIndexHandler;
+    protected InitialContent initialContent;
+    protected SecurityProvider securityProvider;
+    protected IndexEditorProvider indexEditorProvider;
+    protected QueryIndexProvider queryIndexProvider;
+    protected QueryEngineSettings queryEngineSettings;
+
+
+    public TestRepositoryBuilder() {
+        this.isAsync = isAsync;
```
Comment (fabriziofortino) [inline oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/TestRepositoryBuilder.java]: we are assigning the variable to itself

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticPropertyIndexTest1.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Override
+    protected ContentRepository createRepository() {
+        repositoryOptionsUtil = new ElasticTestRepositoryBuilder(elasticRule).build();
+        return repositoryOptionsUtil.getOak().createContentRepository();
+    }
+
+    @Override
+    protected void createTestIndexNode() throws Exception {
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticPropertyIndexTest1.java]:
```suggestion
    protected void createTestIndexNode() {
```

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/LucenePropertyIndexTest1.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Override
+    protected ContentRepository createRepository() {
+        repositoryOptionsUtil = new LuceneTestRepositoryBuilder(executorService, temporaryFolder).build();
+        indexOptions = new LuceneIndexOptions();
+        return repositoryOptionsUtil.getOak().createContentRepository();
+    }
+
+    @Override
+    protected void createTestIndexNode() throws Exception {
```
Comment (fabriziofortino) [inline oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/LucenePropertyIndexTest1.java]:
```suggestion
    protected void createTestIndexNode() {
```

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/ElasticTestRepositoryBuilder.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+public class ElasticTestRepositoryBuilder extends TestRepositoryBuilder {
+
+    private final ElasticConnection esConnection;
+    private final int asyncIndexingTimeInSeconds = 5;
+
+    public ElasticTestRepositoryBuilder(ElasticConnectionRule elasticRule) {
+        this.esConnection = elasticRule.useDocker() ? elasticRule.getElasticConnectionForDocker() :
+                elasticRule.getElasticConnectionFromString();
+        this.editorProvider = (ElasticIndexEditorProvider) getIndexEditorProvider();
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/ElasticTestRepositoryBuilder.java]:
```suggestion
        this.editorProvider = getIndexEditorProvider();
```

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/ElasticTestRepositoryBuilder.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
+package org.apache.jackrabbit.oak.plugins.index;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.jackrabbit.oak.Oak;
+import org.apache.jackrabbit.oak.plugins.index.counter.NodeCounterEditorProvider;
+import org.apache.jackrabbit.oak.plugins.index.elastic.ElasticConnection;
+import org.apache.jackrabbit.oak.plugins.index.elastic.ElasticConnectionRule;
+import org.apache.jackrabbit.oak.plugins.index.elastic.ElasticIndexDefinition;
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/ElasticTestRepositoryBuilder.java]:
```suggestion
```

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/SpellcheckTest1.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+    public void testSpellcheckMultipleWords() throws Exception {
+        adminSession.save();
+        QueryManager qm = adminSession.getWorkspace().getQueryManager();
+        Node par = allow(getOrCreateByPath("/parent", "oak:Unstructured", adminSession));
+        Node n1 = par.addNode("node1");
+        n1.setProperty("foo", "it is always a good idea to go visiting ontario");
+        Node n2 = par.addNode("node2");
+        n2.setProperty("foo", "ontario is a nice place to live in");
+        Node n3 = par.addNode("node3");
+        n2.setProperty("foo", "I flied to ontario for voting for the major polls");
```
Comment (fabriziofortino) [inline oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/SpellcheckTest1.java]: should this be `n3`?

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/SpellcheckTest1.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+        Node n4 = par.addNode("node4");
+        n2.setProperty("foo", "I will go voting in ontario, I always voted since I've been allowed to");
```
Comment (fabriziofortino) [inline oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/SpellcheckTest1.java]: should this be `n4`?

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/SpellcheckTest1.java [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    private Node allow(Node node) throws RepositoryException {
+        AccessControlUtils.allow(node, "anonymous", Privilege.JCR_READ);
+        return node;
+    }
+
+    static List<String> getResult(QueryResult result, String propertyName) throws RepositoryException {
+        List<String> results = Lists.newArrayList();
+        RowIterator it = null;
```
Comment (fabriziofortino) [inline oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/SpellcheckTest1.java]:
```suggestion
        RowIterator it;
```

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/SpellcheckTest1.java [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+import java.util.UUID;
+
+import static org.apache.jackrabbit.commons.JcrUtils.getOrCreateByPath;
+import static org.apache.jackrabbit.oak.plugins.index.search.FulltextIndexConstants.PROP_ANALYZED;
+import static org.apache.jackrabbit.oak.plugins.index.search.FulltextIndexConstants.PROP_USE_IN_SPELLCHECK;
+import static org.junit.Assert.assertEquals;
+
+public abstract class SpellcheckTest1 extends AbstractJcrTest {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SpellcheckTest1.class);
```
Comment (fabriziofortino) [inline oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/SpellcheckTest1.java]: this is never used

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneSpellcheckTest1.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+@Ignore
+public class LuceneSpellcheckTest1 extends SpellcheckTest1 {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LuceneSpellcheckTest1.class);
+
+    private ExecutorService executorService = Executors.newFixedThreadPool(2);
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder(new File("target"));
+
+    protected Repository createJcrRepository() throws RepositoryException {
```
Comment (fabriziofortino) [inline oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneSpellcheckTest1.java]:
```suggestion
    protected Repository createJcrRepository() {
```

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticSpellcheckTest1.java [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.jcr.Repository;
+import javax.jcr.RepositoryException;
+import java.io.IOException;
+
+public class ElasticSpellcheckTest1 extends SpellcheckTest1 {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ElasticSpellcheckTest1.class);
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticSpellcheckTest1.java]: never used

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticSpellcheckTest1.java [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+   Close the ES connection after every test method execution
+    */
+    @After
+    public void cleanup() throws IOException {
+        anonymousSession.logout();
+        adminSession.logout();
+        elasticRule.closeElasticConnection();
+    }
+
+    protected Repository createJcrRepository() throws RepositoryException {
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticSpellcheckTest1.java]:
```suggestion
    protected Repository createJcrRepository() {
```

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneFacetTest1.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+import javax.jcr.Repository;
+import javax.jcr.RepositoryException;
+import java.io.File;
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+public class LuceneFacetTest1 extends FacetTest1 {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LuceneFacetTest1.class);
```
Comment (fabriziofortino) [inline oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneFacetTest1.java]: never used

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneFacetTest1.java [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+    private ExecutorService executorService = Executors.newFixedThreadPool(2);
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder(new File("target"));
+
+    protected Repository createJcrRepository() throws RepositoryException {
```
Comment (fabriziofortino) [inline oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneFacetTest1.java]:
```suggestion
    protected Repository createJcrRepository() {
```

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticFacetTest1.java [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+    
+    /*
+    Close the ES connection after every test method execution
+     */
+    @After
+    public void cleanup() throws IOException {
+        elasticRule.closeElasticConnection();
+    }
+
+    protected Repository createJcrRepository() throws RepositoryException {
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticFacetTest1.java]:
```suggestion
    protected Repository createJcrRepository() {
```

**Review by fabriziofortino — COMMENTED** [review]
I left some comments/questions. 

Overall, I think it's a good job. Only one question: I am able to execute the tests that extend from oak-search from IntelliJ without any problem. But if I run `mvn clean install` they are not executed. Any idea?

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/FacetTest1.java (hunk 3 above)
Comment (tihom88) [reply oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/FacetTest1.java]: Actually these are ported tests with same name. I had add some identifier to differentiate but still keep track of from which test it was ported. As we port keep porting other tests, we can remove the original tests.

**Comment by tihom88** [comment]
> I left some comments/questions.
> 
> Overall, I think it's a good job. Only one question: I am able to execute the tests that extend from oak-search from IntelliJ without any problem. But if I run `mvn clean install` they are not executed. Any idea?

Didn't saw this. Will have a look at why tests are not running with mvn clean install

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/TestRepository.java (hunk 6 above)
Comment (tihom88) [reply oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/TestRepository.java:26]: in some tests we explicitly call asyncIndexUpdate.run. Current ported tests don't have such logic. So yes we can remove this now and reintroduce when we port tests which use asyncIndxUpdate instance.

File: oak-lucene/pom.xml [hunk 22]
```diff
@@ -358,6 +358,13 @@
       <version>${jackrabbit.version}</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.jackrabbit</groupId>
+      <artifactId>oak-search</artifactId>
+      <version>${project.version}</version>
+      <type>test-jar</type>
```
Comment (nit0906) [inline oak-lucene/pom.xml:365]: Is this needed ? Shouldn't scope=test be sufficient ?

File: oak-search-elastic/pom.xml [hunk 23]
```diff
@@ -252,6 +258,14 @@
...
       <artifactId>oak-jcr</artifactId>
       <version>${project.version}</version>
+      <classifier>tests</classifier>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.jackrabbit</groupId>
+      <artifactId>oak-search</artifactId>
+      <version>${project.version}</version>
+      <type>test-jar</type>
```
Comment (nit0906) [inline oak-search-elastic/pom.xml:268]: Same as lucene - is test-jar needed here ?

File: oak-search-elastic/pom.xml [hunk 24]
```diff
@@ -187,6 +187,12 @@
       <version>${project.version}</version>
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.jackrabbit</groupId>
+      <artifactId>oak-jcr</artifactId>
+      <version>${project.version}</version>
+      <scope>provided</scope>
```
Comment (nit0906) [inline oak-search-elastic/pom.xml]: If we added this dep only for the test classes in oak-jcr, then we should limit the scope to test

**Review by nit0906 — COMMENTED** [review]
Added a couple of comments, plus one general question/ point to discuss,

In the elastic test, could we have another layer of abstraction where we can define stuff as the class rule and assert eventually (because these would be exactly same for all the test classes and currently we are writing them in all of them).

We can have a structure something like - 
AbstractBaseTest -> in oak-search (with util/helper stuff common to both elastic and lucene)
AbstractElasticBase - > in oak-search-elastic (with the elastic rule, overriding assertEventually from the above class, stuff like that) that extends the AbstractBaseTest  class

We can make the abstract classes like SpellcheckTest1 as interfaces with default methods (I am not sure if that would work with @Test methods, but we can give it a try) and then the 
final test class in Elastic module would be something like

Elastic<Feature>Test extends AbstractElasticBase implements FeatureTest

File: oak-lucene/pom.xml (hunk 22 above)
Comment (tihom88) [reply oak-lucene/pom.xml:365]: each module in oak build src and test jar. Here oak-search's test jar is required  as abstract tests are written in oak-search-test jar and hence type test-jar . In this pom itself oak-search(src) dependency is also included above.

File: oak-search-elastic/pom.xml (hunk 23 above)
Comment (tihom88) [reply oak-search-elastic/pom.xml:268]: same.

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/SpellcheckTest1.java (hunk 13 above)
Comment (tihom88) [reply oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/SpellcheckTest1.java]: done

File: oak-search-elastic/pom.xml (hunk 24 above)
Comment (tihom88) [reply oak-search-elastic/pom.xml]: done

**Comment by tihom88** [comment]
> I left some comments/questions.
> 
> Overall, I think it's a good job. Only one question: I am able to execute the tests that extend from oak-search from IntelliJ without any problem. But if I run `mvn clean install` they are not executed. Any idea?

Found the issue, surefire have a regex with "*Test" which led to no picking Test1 for execution when we do mvn clean install --> changed name to CommonTest instead of Test1

**Comment by tihom88** [comment]
> Added a couple of comments, plus one general question/ point to discuss,
> 
> In the elastic test, could we have another layer of abstraction where we can define stuff as the class rule and assert eventually (because these would be exactly same for all the test classes and currently we are writing them in all of them).
> 
> We can have a structure something like -
> AbstractBaseTest -> in oak-search (with util/helper stuff common to both elastic and lucene)
> AbstractElasticBase - > in oak-search-elastic (with the elastic rule, overriding assertEventually from the above class, stuff like that) that extends the AbstractBaseTest class
> 
> We can make the abstract classes like SpellcheckTest1 as interfaces with default methods (I am not sure if that would work with @test methods, but we can give it a try) and then the
> final test class in Elastic module would be something like
> 
> ElasticTest extends AbstractElasticBase implements FeatureTest

Adding test as default methods in interface doesn't feel right to me. But lets discuss this.

**Pull request closed without merging**